src/chainstate/stacks/db/transactions.rs

```diff
@@ -8417,4 +8530,286 @@ pub mod test {
+            _check_error,
+        ))) = err
+        {
+        } else {
+            panic!("Did not get unchecked interpreter error");
+        }
+        conn.commit_block();
+
+        // in 2.05, this invalidates the block
+        let mut conn = chainstate.block_begin(
+            &TestBurnStateDB_20,
+            &FIRST_BURNCHAIN_CONSENSUS_HASH,
+            &FIRST_STACKS_BLOCK_HASH,
+            &ConsensusHash([2u8; 20]),
+            &BlockHeaderHash([2u8; 32]),
+        );
+
+        let (fee, _) = StacksChainState::process_transaction(
+            &mut conn,
+            &signed_runtime_checkerror_trait_tx,
+            false,
+        )
+        .unwrap();
+        assert_eq!(fee, 0);
+
+        let (fee, _) = StacksChainState::process_transaction(
+            &mut conn,
+            &signed_runtime_checkerror_impl_tx,
+            false,
+        )
+        .unwrap();
+        assert_eq!(fee, 0);
+
+        let (fee, _) =
+            StacksChainState::process_transaction(&mut conn, &signed_runtime_checkerror_tx, false)
+                .unwrap();
+        assert_eq!(fee, 0);
+
+        let err = StacksChainState::process_transaction(
+            &mut conn,
+            &signed_test_trait_checkerror_tx,
+            false,
+        )
+        .unwrap_err();
+        if let Error::ClarityError(clarity_error::Interpreter(InterpreterError::Unchecked(
+            _check_error,
+        ))) = err
+        {
+        } else {
+            panic!("Did not get unchecked interpreter error");
+        }
+        conn.commit_block();
+
+        // in 2.1, this is a runtime error
+        let mut conn = chainstate.block_begin(
+            &TestBurnStateDB_21,
+            &FIRST_BURNCHAIN_CONSENSUS_HASH,
+            &FIRST_STACKS_BLOCK_HASH,
+            &ConsensusHash([3u8; 20]),
+            &BlockHeaderHash([3u8; 32]),
+        );
+
+        let (fee, _) = StacksChainState::process_transaction(
+            &mut conn,
+            &signed_runtime_checkerror_trait_tx,
+            false,
+        )
+        .unwrap();
+        assert_eq!(fee, 0);
+
+        let (fee, _) = StacksChainState::process_transaction(
+            &mut conn,
+            &signed_runtime_checkerror_impl_tx,
+            false,
+        )
+        .unwrap();
+        assert_eq!(fee, 0);
+
+        let (fee, _) =
+            StacksChainState::process_transaction(&mut conn, &signed_runtime_checkerror_tx, false)
+                .unwrap();
+        assert_eq!(fee, 0);
+
+        let (fee, tx_receipt) = StacksChainState::process_transaction(
+            &mut conn,
+            &signed_test_trait_checkerror_tx,
+            false,
+        )
+        .unwrap();
+        assert_eq!(fee, 0);
+
+        assert!(tx_receipt.vm_error.is_some());
+        let err_str = tx_receipt.vm_error.unwrap();
+        assert!(err_str
+            .find("TypeValueError(OptionalType(TraitReferenceType(TraitIdentifier ")
+            .is_some());
```

Comment:
Can you add checks to this test for the following:

1. In 2.1, the fee (which should be positive for the test to actually assert a change) is assessed against the payer's account, and the nonce is bumped for the failed transaction.
2. In 2.05, the fee isn't assessed (again, it needs to be positive) against the payer's account, and the nonce isn't bumped for the failed transaction.
3. In 2.1, the transaction is committed, but there's no state change. To do this, I think you'd need to invoke `(flip)` at some point in the `test` method before triggering the error. Then just assert that `flip` hasn't changed after the tx.

Then, I think this needs an additional (very similar) battery of tests for the case where the "handled" error comes from a contract publish -- I think if you add a `(test .foo-impl)` invocation to the bottom of the contract publish, it'll create such a situation.